amethyst_renderer/src/tex_animation.rs

```diff
@@ -0,0 +1,178 @@
+    pub fn new(playback_speed: f32, sprite_sheet_data: SpriteSheetDataHandle, sheet_storage: &AssetStorage<SpriteSheetData>, starting_animation: &str) -> Option<SpriteSheetAnimation> {
+        sheet_storage.get(&sprite_sheet_data).and_then(|data| data.animation_mapping.get(starting_animation).map(|i| *i)).map(|current_animation| {
+            SpriteSheetAnimation {
+                playback_speed,
+                sprite_sheet_data,
+                current_animation,
+                current_frame: 0,
+                frame_timer: Duration::from_secs(0),
+            }
+        })
+
+    }
+
+    /// Sets the animation to the named case-sensitive animation from the `SpriteSheetData`.
+    ///
+    /// ## Returns
+    ///
+    /// If the name provided can't be found in the `SpriteSheetData` for this component then this
+    /// will return `Err` otherwise it will return `Ok`.
+    pub fn set_animation(&mut self, sheet_storage: &AssetStorage<SpriteSheetData>, animation: &str) -> Result<(), ()> {
+        let animation = sheet_storage.get(&self.sprite_sheet_data).and_then(|data| data.animation_mapping.get(animation).map(|i| *i));
+        match animation {
+            Some(animation) => {
+                self.current_animation = animation;
+                self.current_frame = 0;
+                self.frame_timer = Duration::from_secs(0);
+                Ok(())
+            }
+            None => {
+                Err(())
+            }
+        }
+    }
+
+    /// Retrieve frame information for the current frame.
+    ///
+    /// ## Returns
+    ///
+    /// Most of the time this should return `Some`.  If it doesn't then either the
+    /// `SpriteSheetDataHandle` used in initialization or the `sheet_storage` in this function are
+    /// invalid.  Because handles take at least one frame to become valid after being loaded the
+    /// handle is the most likely point of failure.
+    pub fn current_frame(&self, sheet_storage: &AssetStorage<SpriteSheetData>) -> Option<Frame> {
+        sheet_storage.get(&self.sprite_sheet_data).map(|data| data.frames[data.animations[self.current_animation][self.current_frame]].clone())
+    }
+
+    /// Advance the current animation by the time given.  Normally this will be called by the
+    /// engine by default using the regularly running game clock.  This function is made public so
+    /// that you can manually advance the animation if you require.
+    pub fn advance(&mut self, time: Duration, sheet_storage: &AssetStorage<SpriteSheetData>) -> Result<(), ()> {
+        use amethyst_core::timing::duration_f32_mul;
+        self.frame_timer += duration_f32_mul(time, self.playback_speed);
+        let mut frame_time = self.current_frame(sheet_storage).ok_or(())?.duration;
+        while self.frame_timer >= frame_time {
+            self.current_frame += 1;
+            let max_frame = sheet_storage.get(&self.sprite_sheet_data).map(|data| data.animations[self.current_animation].len() - 1).ok_or(())?;
```

Comment:
Don't see any chance for failure here. Should be `unwrap`.
Or get `&SpriteSheetData` at the first line and then reuse it.